### PR TITLE
Change json output of exjob containers from stdout to a file

### PR DESCRIPTION
**Pull request by rohitsakala**
[#168510894](https://www.pivotaltracker.com/n/projects/2192232/stories/168510894)

### CODE REVIEW

File: pkg/kube/controllers/extendedjob/job_creator.go [hunk 1]
```diff
@@ -46,9 +56,97 @@ type jobCreatorImpl struct {
...
+		Subjects: []v1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      serviceAccountName,
+				Namespace: namespace,
+			},
+		},
+		RoleRef: v1.RoleRef{
+			Kind:     "ClusterRole",
+			Name:     "cluster-admin",
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_creator.go]: This is too much for this - I believe we can restrict this further.
https://www.pivotaltracker.com/story/show/168884844

Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_creator.go]: This is too much, I believe we can restrict it further.
https://www.pivotaltracker.com/story/show/168884844

File: e2e/cli/util_bpm_configs_test.go [hunk 2]
```diff
@@ -35,9 +36,11 @@ var _ = Describe("bpm-configs", func() {
 
 			Eventually(session).Should(gexec.Exit(0))
 
-			output := session.Out.Contents()
-			jsonOutput := map[string]string{}
-			err = json.Unmarshal(output, &jsonOutput)
+			var jsonOutput map[string]string
+			dataBytes, err := ioutil.ReadFile(assetPath + "/output.json")
```
Comment (viovanov) [inline e2e/cli/util_bpm_configs_test.go]: please use filepath.Join

File: pkg/bosh/manifest/cmd_interpolate_test.go [hunk 3]
```diff
@@ -31,40 +31,25 @@ instance_groups:
 - name: ((value2.key3))
 `)
 		varDir = assetPath + "/vars"
+		outputFilePath = assetPath + "/output.json"
```
Comment (viovanov) [inline pkg/bosh/manifest/cmd_interpolate_test.go]: filepath.Join

File: pkg/kube/controllers/extendedjob/errand_reconciler.go [hunk 4]
```diff
@@ -34,6 +34,8 @@ const (
 	EnvCfOperatorAzIndex = "AZ_INDEX"
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/errand_reconciler.go]: Can you please rename this to `EnvCFOperatorAZIndex`?

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+	"os"
+	"regexp"
+	"strconv"
+	"time"
+
+	"code.cloudfoundry.org/cf-operator/pkg/bosh/converter"
+	ejv1 "code.cloudfoundry.org/cf-operator/pkg/kube/apis/extendedjob/v1alpha1"
+	"code.cloudfoundry.org/cf-operator/pkg/kube/client/clientset/versioned"
+	"code.cloudfoundry.org/cf-operator/pkg/kube/util"
+	podutil "code.cloudfoundry.org/cf-operator/pkg/kube/util/pod"
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: please split operator imports from third-party imports

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+)
+
+// PersistOutput converts the output files of each container
+// in the pod related to an ejob into a kubernetes secret.
+func PersistOutput(namespace string) error {
+
+	// hostname of the container is the pod name in kubernetes
+	podName, err := os.Hostname()
+	if err != nil {
+		return errors.Wrapf(err, "Failed to fetch pod name.")
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: no uppercase on error messages

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+	}
+	if podName == "" {
+		return errors.Wrapf(err, "Pod name is empty.")
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: no uppercase on error messages

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+	errorChannel := make(chan error)
+
+	// Loop over containers and create secrets for each output file for container
+	for containerIndex, container := range pod.Spec.Containers {
+
+		if container.Name == "output-persist" {
+			continue
+		}
+
+		filePath := fmt.Sprintf("%s%s%s", "/mnt/quarks/", container.Name, "/output.json")
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: please use `filepath.Join`

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+	for i := 0; i < len(pod.Spec.Containers)-1; i++ {
+		select {
+		case containerIndex := <-fileNotifyChannel:
+			outputContainer := pod.Spec.Containers[containerIndex]
+			err := createOutputSecret(outputContainer, namePrefix, namespace, pod.Name, clientSet, exjob)
+			if err != nil {
+				return err
+			}
+		case failure := <-errorChannel:
+			return errors.Wrapf(failure, "Failure waiting for output file for container in pod %s", pod.GetName())
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: no uppercase on error messages

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+		}
+	}
+	return nil
+}
+
+// waitForFile waits for the file to be created
+func waitForFile(containerIndex int, filePath string, fileNotifyChannel chan<- int, errorChannel chan<- error) {
+
+	for {
+		time.Sleep(5 * time.Second)
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: we should use https://github.com/fsnotify/fsnotify instead

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+
+	return clientSet, versionedClientSet, nil
+}
+
+func createOutputSecret(outputContainer corev1.Container, namePrefix string, namespace string, podName string, clientSet *kubernetes.Clientset, exjob *ejv1.ExtendedJob) error {
+
+	secretName := namePrefix + outputContainer.Name
+
+	// Fetch json from file
+	filePath := fmt.Sprintf("%s%s%s", "/mnt/quarks/", outputContainer.Name, "/output.json")
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: filepath.Join

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+	file, _ := ioutil.ReadFile(filePath)
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: ❗️  you're ignoring an error here

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+
+		if err != nil {
+			if apierrors.IsAlreadyExists(err) {
+				// If it exists update it
+				_, err = clientSet.CoreV1().Secrets(namespace).Update(secret)
+				if err != nil {
+					return errors.Wrapf(err, "Failed to update secret %s for container %s in pod %s.", secretName, outputContainer.Name, podName)
+				}
+			} else {
+				return errors.Wrapf(err, "Failed to create secret %s for container %s in pod %s.", secretName, outputContainer.Name, podName)
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: shouldn't capitalize first letter in an error

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+		secret.Labels = secretLabels
+
+		_, err = clientSet.CoreV1().Secrets(namespace).Create(secret)
+
+		if err != nil {
+			if apierrors.IsAlreadyExists(err) {
+				// If it exists update it
+				_, err = clientSet.CoreV1().Secrets(namespace).Update(secret)
+				if err != nil {
+					return errors.Wrapf(err, "Failed to update secret %s for container %s in pod %s.", secretName, outputContainer.Name, podName)
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: shouldn't capitalize first letter in an error

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+			} else {
+				return errors.Wrapf(err, "Failed to create secret %s for container %s in pod %s.", secretName, outputContainer.Name, podName)
+			}
+		}
+
+	}
+	return nil
+}
+
+func createVersionSecret(clientSet *kubernetes.Clientset, namespace string, ownerName string, ownerID types.UID, secretName string, secretData map[string]string, labels map[string]string, sourceDescription string) error {
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: we already have a package that deals with listing and reading versioned secrets - can we use that?

File: pkg/kube/controllers/extendedjob/job_creator.go [hunk 16]
```diff
@@ -46,9 +56,97 @@ type jobCreatorImpl struct {
...
+		},
+	}
+
+	// Set serviceaccount to the pod
+	template.Spec.ServiceAccountName = serviceAccountName
+
+	err = j.client.Create(ctx, serviceAccount)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return false, errors.Wrapf(err, "Could not create service account for pod in ejob %s.", eJob.Name)
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_creator.go]: shouldn't capitalize first letter in an error

File: pkg/kube/controllers/extendedjob/job_creator.go [hunk 17]
```diff
@@ -46,9 +56,97 @@ type jobCreatorImpl struct {
...
+		}
+	}
+
+	err = j.client.Create(ctx, roleBinding)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return false, errors.Wrapf(err, "Could not create role binding for pod in ejob '%s'", eJob.Name)
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_creator.go]: shouldn't capitalize first letter in an error

File: pkg/kube/controllers/extendedjob/job_creator.go [hunk 18]
```diff
@@ -46,9 +56,97 @@ type jobCreatorImpl struct {
...
+			},
+		},
+	}
+
+	// Loop through containers and add quarks logging volume specs.
+	for containerIndex, container := range template.Spec.Containers {
+
+		// Add pod volume specs to the pod
+		podVolumeSpec := corev1.Volume{
+			Name:         fmt.Sprintf("%s%s", "output-", container.Name),
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_creator.go]: you should add a function in our naming package for this

File: pkg/kube/controllers/extendedjob/job_creator.go [hunk 19]
```diff
@@ -46,9 +56,97 @@ type jobCreatorImpl struct {
...
+			VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}},
+		}
+		template.Spec.Volumes = append(template.Spec.Volumes, podVolumeSpec)
+
+		// Add container volume specs to continer
+		containerVolumeMountSpec := corev1.VolumeMount{
+			Name:      fmt.Sprintf("%s%s", "output-", container.Name),
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_creator.go]: please use a naming package

File: pkg/kube/controllers/extendedjob/job_creator.go [hunk 20]
```diff
@@ -46,9 +56,97 @@ type jobCreatorImpl struct {
...
+			MountPath: mountPath,
+		}
+		template.Spec.Containers[containerIndex].VolumeMounts = append(template.Spec.Containers[containerIndex].VolumeMounts, containerVolumeMountSpec)
+
+		// Add container volume spec to output persist container
+		containerVolumeMountSpec.MountPath = fmt.Sprintf("%s%s", mountPath, container.Name)
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_creator.go]: filepath.Join

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+	}
+
+	// Fetch the pod
+	pod, err := clientSet.CoreV1().Pods(namespace).Get(podName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to fetch pod %s", podName)
+	}
+
+	// Fetch the exjob
+	exjobName := pod.GetLabels()["fissile.cloudfoundry.org/ejob-name"]
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: please use a constant

File: pkg/kube/controllers/extendedjob/job_create_secret.go [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generatedSecretName,
+			Namespace: namespace,
+			Labels:    labels,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         "fissile.cloudfoundry.org/v1alpha1",
```
Comment (viovanov) [inline pkg/kube/controllers/extendedjob/job_create_secret.go]: please use a constant - we already define this

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 15 above)
Comment (manno) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: That package uses the controller-runtime client.

File: pkg/kube/controllers/extendedjob/job_creator.go (hunk 1 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_creator.go]: Thanks. Sure.

File: cmd/internal/persist_output.go [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+
+const (
+	persistOutputFileFailedMessage = "persist-output command failed."
+)
+
+// persistOutputCmd is the persist-output command.
+var persistOutputCmd = &cobra.Command{
+	Use:   "persist-output [flags]",
+	Short: "Persist a file into a kube secret",
+	Long: `Persists a log file created  by containers in a pod of extendedjob
```
[the remark this answers is not in the document]
Comment (rohitsakala) [reply cmd/internal/persist_output.go]: Done.

File: cmd/internal/persist_output.go [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"code.cloudfoundry.org/cf-operator/pkg/kube/controllers/extendedjob"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+const (
+	persistOutputFileFailedMessage = "persist-output command failed."
```
[the remark this answers is not in the document]
Comment (rohitsakala) [reply cmd/internal/persist_output.go]: removed.

File: e2e/cli/util_bpm_configs_test.go (hunk 2 above)
Comment (rohitsakala) [reply e2e/cli/util_bpm_configs_test.go]: Done.

File: pkg/kube/controllers/extendedjob/errand_reconciler.go (hunk 4 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/errand_reconciler.go]: Done.

File: pkg/bosh/manifest/cmd_interpolate_test.go (hunk 3 above)
Comment (rohitsakala) [reply pkg/bosh/manifest/cmd_interpolate_test.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 5 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 6 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 7 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 9 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 8 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 10 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 11 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 13 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 12 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Cared.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 14 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

File: pkg/kube/controllers/extendedjob/job_creator.go (hunk 16 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_creator.go]: Done.

File: pkg/kube/controllers/extendedjob/job_creator.go (hunk 17 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_creator.go]: Done.

File: pkg/kube/controllers/extendedjob/job_creator.go (hunk 18 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_creator.go]: Used Sanitize func

File: pkg/kube/controllers/extendedjob/job_creator.go (hunk 19 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_creator.go]: Used Sanitize func

File: pkg/kube/controllers/extendedjob/job_creator.go (hunk 20 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_creator.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 21 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

File: pkg/kube/controllers/extendedjob/job_create_secret.go (hunk 22 above)
Comment (rohitsakala) [reply pkg/kube/controllers/extendedjob/job_create_secret.go]: Done.

**Pull request merged**